// angular-formio/index.js
/**
 * @fileoverview added by tsickle
 * Generated from: index.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export { FormioAppConfig } from './formio.config';
export { FormioError } from './formio.common';
export { FormioService } from './formio.service';
export { extendRouter } from './formio.utils';
export { FormioBaseComponent } from './FormioBaseComponent';
export { FormioComponent } from './components/formio/formio.component';
export { FormBuilderComponent } from './components/formbuilder/formbuilder.component';
export { FormioLoader } from './components/loader/formio.loader';
export { FormioLoaderComponent } from './components/loader/formio.loader.component';
export { FormioAlerts } from './components/alerts/formio.alerts';
export { FormioAlertsComponent } from './components/alerts/formio.alerts.component';
export { createCustomFormioComponent } from './custom-component/create-custom-component';
export { registerCustomTag, registerCustomTags, registerCustomFormioComponent, registerCustomFormioComponentWithClass } from './custom-component/register-custom-component';
export { FormioModule } from './formio.module';
export { Formio, Utils as FormioUtils, Templates, Components } from 'formiojs';
